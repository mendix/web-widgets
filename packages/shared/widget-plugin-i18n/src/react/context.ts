import { createContext } from "react";
import { PluginI18n } from "../types";

const contextPath = "com.mendix.widget.plugin.i18n.context" as const;

type I18nContext = React.Context<PluginI18n | null>;

const createI18nContext = (): I18nContext => createContext<PluginI18n | null>(null);

declare global {
    interface Window {
        [contextPath]?: ReturnType<typeof createI18nContext>;
    }
}

export function getGlobalI18nPluginContext(): I18nContext {
    return window[contextPath] ?? createI18nContext();
}
