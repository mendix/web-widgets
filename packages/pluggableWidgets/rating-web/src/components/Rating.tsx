import { CSSProperties, ReactElement, useRef, useState } from "react";
import classNames from "classnames";

export interface RatingProps {
    animated: boolean;
    className?: string;
    disabled: boolean;
    emptyIcon: ReactElement;
    fullIcon: ReactElement;
    maximumValue: number;
    onChange?: (value: number) => void;
    style?: CSSProperties;
    tabIndex?: number;
    value: number;
}

const enum Direction {
    PREVIOUS = "previous",
    NEXT = "next"
}

export function Rating(props: RatingProps): ReactElement {
    const [hover, setHover] = useState<undefined | number>(undefined);
    const containerRef = useRef<HTMLDivElement>(null);
    const onClickAction = !props.disabled
        ? (currentIndex: number): void => {
              if (props.value === currentIndex) {
                  props.onChange?.(0);
              } else {
                  props.onChange?.(currentIndex);
              }
          }
        : undefined;

    const focusItem = (direction: Direction): void => {
        if (containerRef.current) {
            const currentFocusedElement = containerRef.current.querySelector(".rating-item:focus");
            if (direction === Direction.NEXT && currentFocusedElement?.nextSibling) {
                (currentFocusedElement.nextSibling as HTMLDivElement).focus();
            } else if (direction === Direction.PREVIOUS && currentFocusedElement?.previousSibling) {
                (currentFocusedElement.previousSibling as HTMLDivElement).focus();
            }
        }
    };

    return (
        <div
            className={classNames("widget-rating", props.className)}
            data-focusindex={props.tabIndex ?? 0}
            role="radiogroup"
            style={props.style}
            onKeyDown={event => {
                switch (event.key) {
                    case "Left": // Microsoft Edge value
                    case "ArrowLeft":
                        event.preventDefault();
                        focusItem(Direction.PREVIOUS);
                        break;
                    case "Right": // Microsoft Edge value
                    case "ArrowRight":
                        event.preventDefault();
                        focusItem(Direction.NEXT);
                        break;
                }
            }}
            ref={containerRef}
        >
            {Array.from({ length: props.maximumValue }, (_, index) => {
                const currentIndex = index + 1;
                return (
                    <div
                        aria-checked={currentIndex === props.value}
                        aria-label={currentIndex.toString()}
                        className={classNames("rating-item", { disabled: props.disabled })}
                        key={`icon_${currentIndex}`}
                        onClick={() => {
                            setHover(undefined);
                            onClickAction?.(currentIndex);
                        }}
                        onKeyDown={event => {
                            if (event.key === " " || event.key === "Enter") {
                                event.preventDefault();
                                onClickAction?.(currentIndex);
                            }
                        }}
                        onMouseOut={() => {
                            setHover(undefined);
                        }}
                        onMouseOver={() => {
                            if (!props.disabled && props.animated) {
                                setHover(currentIndex);
                            }
                        }}
                        role="radio"
                        tabIndex={index === Math.max(props.value - 1, 0) ? 0 : -1}
                    >
                        {Number(currentIndex) <= props.value ? (
                            props.fullIcon
                        ) : hover && currentIndex <= Number(hover) ? (
                            <div className="rating-item-hover">{props.fullIcon}</div>
                        ) : (
                            props.emptyIcon
                        )}
                    </div>
                );
            })}
        </div>
    );
}
