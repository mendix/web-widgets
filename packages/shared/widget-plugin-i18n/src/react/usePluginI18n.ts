import { useState } from "react";
import { I18nController } from "../controllers/I18nController";
import { PluginI18n } from "../types";

export function usePluginI18n(languageCode: string): PluginI18n | null {
    const [plugin, setPlugin] = useState<PluginI18n | null>(() => {
        new I18nController({
            languageCode,
            onPluginChange: plugin => setPlugin(plugin)
        });

        return null;
    });

    return plugin;
}
