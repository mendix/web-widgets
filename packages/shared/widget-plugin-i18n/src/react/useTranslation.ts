import { useContext } from "react";
import { PluginI18n } from "../types";
import { getGlobalI18nPluginContext } from "./context";

export function useTranslation(): PluginI18n | null {
    return useContext(getGlobalI18nPluginContext());
}
