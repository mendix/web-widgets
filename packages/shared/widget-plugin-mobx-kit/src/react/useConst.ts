import { useRef } from "react";

export function useConst<T>(fn: T): T extends (...args: never) => infer R ? R : T {
    return (useRef<T | null>(null).current ??= typeof fn === "function" ? fn() : fn);
}
