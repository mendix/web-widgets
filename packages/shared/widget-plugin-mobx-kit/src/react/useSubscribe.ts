import { IReactionOptions, reaction } from "mobx";
import { useEffect, useState } from "react";

export function useSubscribe<Store extends object, T, FireImmediately extends boolean = false>(
    init: Store | (() => Store),
    selector: (store: Store) => T,
    options?: IReactionOptions<T, FireImmediately>
): [T, Store] {
    const [store] = useState(() => (typeof init === "function" ? init() : init));
    const [value, setValue] = useState(() => selector(store));

    function setup(): () => void {
        return reaction(() => selector(store), setValue, options);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(setup, []);

    return [value, store];
}
