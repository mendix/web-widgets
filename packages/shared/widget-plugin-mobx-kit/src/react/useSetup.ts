import { useEffect } from "react";
import { useConst } from "./useConst";

export function useSetup<
    T extends {
        setup(): void | (() => void);
    }
>(fn: () => T): T {
    const obj = useConst(fn);
    useEffect(() => obj.setup(), [obj]);
    return obj;
}
