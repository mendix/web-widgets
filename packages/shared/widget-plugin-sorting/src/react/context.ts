import { useConst } from "@mendix/widget-plugin-mobx-kit/react/useConst";
import { generateUUID } from "@mendix/widget-plugin-platform/framework/generate-uuid";
import { Context, createContext, useContext, useEffect } from "react";
import { SortStoreHost } from "../stores/SortStoreHost";
import { Result, error, value } from "./result-meta";

export interface SortAPI {
    version: 1;
    host: SortStoreHost;
}

const SORT_PATH = "com.mendix.widgets.web.sortable.sortContext";

export function getGlobalSortContext(): Context<SortAPI | null> {
    const scope = window.top === window ? window : window.top;
    return ((scope as any)[SORT_PATH] ??= createContext<SortAPI | null>(null));
}

export function useSortAPI(): Result<SortAPI, Error> {
    const api = useContext(getGlobalSortContext());
    if (api === null) {
        return error(new Error("Error: widget is out of context. Please place the widget inside the Gallery header."));
    }
    return value(api);
}

export function useLockSortAPI(api: SortAPI): Result<SortAPI, Error> {
    const id = useLock(api);

    if (api.host.usedBy !== id) {
        return error(
            new Error(
                `Error: Sort API is already in use by another widget. Remove other sort widgets and refresh the page.`
            )
        );
    }

    return value(api);
}

function useLock({ host }: SortAPI): string {
    const [unlock, id] = useConst(() => {
        const id = `useLock@${generateUUID()}`;
        return [host.lock(id), id] as const;
    });

    useEffect(() => unlock, [unlock]);

    return id;
}
