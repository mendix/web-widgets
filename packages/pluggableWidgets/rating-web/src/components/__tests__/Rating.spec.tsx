import { createElement } from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { Rating, RatingProps } from "../Rating";
import "@testing-library/jest-dom";

const defaultProps: RatingProps = {
    className: "",
    animated: true,
    disabled: false,
    emptyIcon: <div className="empty" />,
    fullIcon: <div className="full" />,
    value: 0,
    onChange: undefined,
    maximumValue: 5,
    style: undefined
};

function renderRating(props?: Partial<RatingProps>) {
    return render(<Rating {...defaultProps} {...props} />);
}

describe("Rating rendering", () => {
    it("renders correctly the structure", () => {
        const { asFragment } = renderRating();
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly the structure when disabled", () => {
        const { asFragment } = renderRating({ disabled: true });
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly the structure without animations", () => {
        const { asFragment } = renderRating({ animated: false });
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly the structure with custom class", () => {
        const { asFragment } = renderRating({ className: "my-custom-class" });
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders the correct amount of items", () => {
        renderRating({ maximumValue: 2 });
        expect(screen.getAllByRole("radio")).toHaveLength(2);
    });

    it("renders the correct amount of items when value is superior to maximumValue", () => {
        const { container } = renderRating({ maximumValue: 2, value: 5 });
        expect(screen.getAllByRole("radio")).toHaveLength(2);
        expect(container.querySelectorAll(".full")).toHaveLength(2);
    });
});

describe("Rating events", () => {
    it("triggers the event with correct value on click", () => {
        const onChange = jest.fn();
        renderRating({ onChange });
        const radios = screen.getAllByRole("radio");
        fireEvent.click(radios[0]);
        expect(onChange).toHaveBeenCalledWith(1);
    });

    it("cleans the value when clicking twice at same value", () => {
        let value = 0;
        const onChange = jest.fn(v => {
            value = v;
        });
        const { rerender } = renderRating({ onChange, value });
        const radios = screen.getAllByRole("radio");
        fireEvent.click(radios[1]);
        expect(onChange).toHaveBeenCalledWith(2);
        rerender(<Rating {...defaultProps} onChange={onChange} value={2} />);
        fireEvent.click(radios[1]);
        expect(onChange).toHaveBeenCalledWith(0);
    });

    it("triggers the event with correct value on space key down", () => {
        const onChange = jest.fn();
        renderRating({ onChange });
        const radios = screen.getAllByRole("radio");
        fireEvent.keyDown(radios[0], { key: " " });
        expect(onChange).toHaveBeenCalledWith(1);
    });

    it("triggers the event with correct value on enter key down", () => {
        const onChange = jest.fn();
        renderRating({ onChange });
        const radios = screen.getAllByRole("radio");
        fireEvent.keyDown(radios[2], { key: "Enter" });
        expect(onChange).toHaveBeenCalledWith(3);
    });

    it("doesn't trigger any event when disabled", () => {
        const onChange = jest.fn();
        renderRating({ disabled: true, onChange });
        const radios = screen.getAllByRole("radio");
        fireEvent.keyDown(radios[1], { key: "Enter" });
        fireEvent.keyDown(radios[2], { key: " " });
        fireEvent.click(radios[3]);
        expect(onChange).not.toHaveBeenCalled();
    });
});
