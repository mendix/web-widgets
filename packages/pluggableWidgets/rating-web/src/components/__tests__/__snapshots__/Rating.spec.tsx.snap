// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating rendering renders correctly the structure 1`] = `
<DocumentFragment>
  <div
    class="widget-rating"
    data-focusindex="0"
    role="radiogroup"
  >
    <div
      aria-checked="false"
      aria-label="1"
      class="rating-item"
      role="radio"
      tabindex="0"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="2"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="3"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="4"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="5"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rating rendering renders correctly the structure when disabled 1`] = `
<DocumentFragment>
  <div
    class="widget-rating"
    data-focusindex="0"
    role="radiogroup"
  >
    <div
      aria-checked="false"
      aria-label="1"
      class="rating-item disabled"
      role="radio"
      tabindex="0"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="2"
      class="rating-item disabled"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="3"
      class="rating-item disabled"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="4"
      class="rating-item disabled"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="5"
      class="rating-item disabled"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rating rendering renders correctly the structure with custom class 1`] = `
<DocumentFragment>
  <div
    class="widget-rating my-custom-class"
    data-focusindex="0"
    role="radiogroup"
  >
    <div
      aria-checked="false"
      aria-label="1"
      class="rating-item"
      role="radio"
      tabindex="0"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="2"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="3"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="4"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="5"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rating rendering renders correctly the structure without animations 1`] = `
<DocumentFragment>
  <div
    class="widget-rating"
    data-focusindex="0"
    role="radiogroup"
  >
    <div
      aria-checked="false"
      aria-label="1"
      class="rating-item"
      role="radio"
      tabindex="0"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="2"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="3"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="4"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
    <div
      aria-checked="false"
      aria-label="5"
      class="rating-item"
      role="radio"
      tabindex="-1"
    >
      <div
        class="empty"
      />
    </div>
  </div>
</DocumentFragment>
`;
